rfcs/032-calm-deletions/README.md

```diff
@@ -0,0 +1,30 @@
+# RFC 032: Calm deletion watcher
+
+**Status:** Draft
+
+**Last updated:** 28/01/2021
+
+## Background
+
+We've realised that the pipeline does not pick up deleted Calm records - once they're deleted there is no trace of them in the Calm API, but they remain in the source VHS tables. Especially because a cleanup project at the end of last year was responsible for a very large number of deletions, this needs to be resolved.
+
+So far as we can tell, there is no way to find deleted records other than checking for their absence. This means that our solution needs to take some sort of polling-based approach: going through the source table and checking for the continued presence of each record in the Calm API.
+
+## Proposal
+
+- The deletion watcher will be an ECS task within the existing calm adapter cluster.
+- It will run as a scheduled task at a frequency to be agreed with the collections information team.
```

Comment:
We wouldn't necessarily have to update all records at the same time, it could be staggered over a week. For example one day we update all records IDs starting with `0-1`, the next day `2-3` etc. That would mean we would be say only making 50,000 requests in a day to the Calm API rather than 300,000.

Reply:
I was going to go down this route but decided not to - what's the advantage of it? Would the Calm API "prefer" it? As I saw it, the average time to notice a deletion stays exactly the same but we suffer having to introduce state

Reply:
Well if we don't want to DOS the Calm API we will artificially limit it in some way (via limiting number of concurrent workers, time in between requests etc), this is just another way of doing so. This suggestion does not introduce state any more than otherwise, it would be hard coded into the worker code to know what range a particular scan is.

But yeah, can see it might not be necessary to do this :shrug:

Reply:
Not necessarily, if you use the reindexer you can get something close to this.

DynamoDB automatically divides the table into segments when you do a Parallel Scan (https://alexwlchan.net/2018/08/parallel-scan-scanamo/). On Monday ask for segment 0 of 7, Tuesday 1 of 7, Wednesday 3 of 7, etc. You won't get contiguous ranges, but since the Calm IDs are UUIDs that's probably fine.

I'm still not sure it's worth doing – I'd rather blast through the table once a week and be done with it – but implementation isn't a bar to doing this if we want to.

Reply:
Doesn't that segment counter state have to live somewhere? Or are you saying the app looks at the clock and decides a segment based on that?

Anyway, I think I would regard this as something useful to come back to if we run into performance/throughput issues.

Reply:
> Doesn't that segment counter state have to live somewhere? Or are you saying the app looks at the clock and decides a segment based on that?

Pretty much – or whatever schedules it does.

> Anyway, I think I would regard this as something useful to come back to if we run into performance/throughput issues.

Agreed.

Reply:
I seem to remember Calm being unhappy about high volumes of requests.

Reply:
> I seem to remember Calm being unhappy about high volumes of requests.

Yup I can't remember all the details but have a memory that ingesting the data into the adaptors initially took a fair amount of manual intervention and lots of things on DLQs

Reply:
:point_up: I think this was made worse by the fact that there are some `ModifiedDate`'s with 10s of 1000s of records. As we are making requests by IDs for this functionality I suspect it will be less of an issue.